README.md
# cgbeacon2

[![Build Status](https://travis-ci.com/Clinical-Genomics/cgbeacon2.svg?branch=master)](https://travis-ci.com/Clinical-Genomics/cgbeacon2)
[![codecov](https://codecov.io/gh/Clinical-Genomics/cgbeacon2/branch/master/graph/badge.svg?token=C7JSM0UKCQ)](https://codecov.io/gh/Clinical-Genomics/cgbeacon2)

An Python and MongoDB - based beacon supporting [ GA4GH API 1.0 ][ga4gh_api1]

This README only gives a brief overview of the tool, for a more complete reference, please check out our docs: https://clinical-genomics.github.io/cgbeacon2

Table of Contents:
1. [ Running the app using Docker ](#docker)
2. [ Installation using pip in a virtual environment](#installation)
3. [ Loading demo data into the database ](#Loading)
4. [ Server endpoints ](#endpoints)
    - [ Obtaining beacon info ](#info)
    - [ Running queries ](#query)
    - [ Adding variants ](#add)
    - [ Removing variants ](#delete)
5. [ Web interface ](#webform)


<a name="docker"></a>
## Installing and running the app using Docker

A demo instance of the app containing test data can be started from the docker-compose file using this command:
```
docker-compose up -d
```
The command will create 3 containers:
- mongodb: starting a mongodb server
- beacon-cli: the a command-line app, which will connect to the server and populates it with demo data
- beacon-web: a web server running on localhost, port 6000.

The server will be running and accepting requests sent from outside the containers (another terminal or a web browser). Read further down to find out about request types and queries.

To stop the containers (and the server), run:
```
docker-compose down
```

To instantiate a server without demo data just remove the line:
`command: bash -c 'beacon add demo'`
from the docker-compose.yml file.

More info on how to set up a server containing app backend and frontend is available in the [docs](https://clinical-genomics.github.io/cgbeacon2)

<a name="installation"></a>
## Installation

### Prerequisites
- It is recommended to install the app inside a virtual environment containing python >3.6
- Bedtools should be installed in the virtual environment. To install it using the bioconda channel type:
```
conda install -n your_beacon_environment -c bioconda bedtools
```
Python 3.6+
- A working instance of **MongoDB**. From the mongo shell you can create a database using this syntax:
```
use cgbeacon2-test
```

Activate the virtual environment and clone this repository from github using this command:
```
git clone https://github.com/Clinical-Genomics/cgbeacon2.git
```

Change directory to the cloned folder and from there, install the software using the following command:
```
pip install -r requirements.txt
pip install -e .
```

To customize the server configuration you'll need to edit the **config.py** file under the /instance folder. &nbsp;
For testing purposes you can keep the default configuration values as they are, but keep in mind that you should adjust these parameters when in production.

To start the server run this command:
```
beacon run -h custom_host -p custom_port
```
Omitting custom_host and custom_port parameters will make the server run on default host=localhost (127.0.0.1) and default port 5000.

Please note that this code is NOT guaranteed to be bug-free and it must be adapted to be used in production.

<a name="loading"></a>
## Loading demo data into the database
In order to test how the software works and run test queries, a demo dataset and variants can be loaded into the database by running the following command:
```
beacon add demo
```
This command will create 2 collections: "dataset" and "variant". Dataset collection will contain a public dataset named "test_public". "variant" collection will be populated with several hundreds variants, both single nucleotide variants, indels and structural variants.

<a name="endpoints"></a>
## Server endpoints

<a name="info"></a>
- **/**.
General info regarding this Beacon, including a description of its datasets, API version, sample count etc, can be obtained by sending a GET request using the following shell command:
```
curl -X GET 'http://localhost:5000/apiv1.0/info'
```

Demo beacon will reply to this request with a JSON object like this:
```
{"alternativeUrl":null,"apiVersion":"v1.0.0","createDateTime":"Tue, 23 Jun 2020 14:33:52 GMT","datasets":[{"assembly_id":"GRCh37","callCount":483,"created":"Tue, 23 Jun 2020 14:33:52 GMT","id":"test_public","info":{"accessType":"PUBLIC"},"name":"Test public dataset","sampleCount":1,"updated":"Tue, 23 Jun 2020 14:33:53 GMT","variantCount":408,"version":1.0}],"description":"Beacon description","id":"SciLifeLab-beacon","name":"SciLifeLab Stockholm Beacon","organisation":{"address":"","contactUrl":"","description":"A science lab","id":"scilifelab","info":[],"logoUrl":"","name":"Clinical Genomics, SciLifeLab","welcomeUrl":""},"sampleAlleleRequests":[{"alternateBases":"A","assemblyId":"GRCh37","datasetIds":["test_public"],"includeDatasetResponses":"HIT","referenceBases":"C","referenceName":"1","start":156146085},{"assemblyId":"GRCh37","includeDatasetResponses":"ALL","referenceBases":"C","referenceName":"20","start":54963148,"variantType":"DUP"}],"updateDateTime":"Tue, 23 Jun 2020 14:33:53 GMT","version":"v1.1","welcomeUrl":null}
```

<a name="query"></a>
- **/query**.
Query endpoint supports both GET and POST requests.
Example of a GET request:
```
curl -X GET \
  'http://localhost:5000/apiv1.0/query?referenceName=1&referenceBases=C&start=156146085&assemblyId=GRCh37&alternateBases=A'
```

Example of a POST request:
```
curl -X POST \
  -H 'Content-Type: application/json' \
  -d '{"referenceName": "1",
  "start": 156146085,
  "referenceBases": "C",
  "alternateBases": "A",
  "assemblyId": "GRCh37",
  "includeDatasetResponses": "HIT"}' http://localhost:5000/apiv1.0/query
```

The Beacon reply to a query of this type would be a json object where the "exist" key will be true if the allele is found, otherwise it will be false.
```
{"allelRequest":{"alternateBases":"A","assemblyId":"GRCh37","datasetIds":[],"includeDatasetResponses":"NONE","referenceBases":"C","referenceName":"1","start":"156146085"},"apiVersion":"1.0.0","beaconId":"SciLifeLab-beacon","datasetAlleleResponses":[],"error":null,"exists":true}
```

<a name="add"></a>
- **/add**.
Example of a valid POST request to the add endpoint:
```
curl -X POST \
  -H 'Content-Type: application/json' \
  -H 'X-Auth-Token: auth_token' \
  -d '{"dataset_id": "test_public",
  "vcf_path": "path/to/cgbeacon2/cgbeacon2/resources/demo/test_trio.vcf.gz",
  "samples" : ["ADM1059A1", "ADM1059A2"],
  "genes" : {"ids": [17284, 29669, 11592], "id_type":"HGNC"},
  "assemblyId": "GRCh37"}' http://localhost:5000/apiv1.0/add
```
Note that only authenticated users will be able to use this endpoint by including the user's auth_token in the request headers. [Additional info](https://clinical-genomics.github.io/cgbeacon2/loading/) on how to use this endpoint.

<a name="delete"></a>
- **/delete**.
Endpoint that can be used by authenticated users to remove variants by sending a DELETE request like this:
```
curl -X DELETE \
  -H 'Content-Type: application/json' \
  -H 'X-Auth-Token: auth_token' \
  -d '{"dataset_id": "test_public",
  "samples" : ["ADM1059A1", "ADM1059A2"]}' http://localhost:5000/apiv1.0/delete
```
Additional info available [here](https://clinical-genomics.github.io/cgbeacon2/removing/).


<a name="webform"></a>
## Web interface
A simple web interface to perform interactive queries can be used by typing the following address in any browser window: `http://127.0.0.1:5000/apiv1.0/query_form`

![Interface picture](docs/pics/beacon2_interface.jpg)

At the moment this interface is disconnected with Elixir AAI so all queries will be limited to the available public datasets in the Beacon.

[ga4gh_api1]: https://github.com/ga4gh-beacon/specification/blob/develop/beacon.md


CHANGELOG.md
## [4.5.1] - 2023-11-14
### Fixed
- Fixed Gunicorn worked not booting by switching to a `python3.8-slim-bullseye` - based image in Dockerfiles

## [4.5] - 2023-11-09
### Changed
- Unfreeze pymongo lib in requirements.txt
- Use MongoDB v7 in docker-compose file
- Force Docker build to use the linux/amd64 architecture in docker-compose file

## [4.4] - 2023-03-17
### Changed
- Updated version of several Docker images used in GitHub actions
### Fixed
- Replaced deprecated Ubuntu version with latest in GitHub actions

## [4.3] - 2022.10.14
### Added
- Improved documentation on how to create an API user.
### Changed
- Code refactoring to eliminate all SonarCloud major and critical issues
### Fixed
- Potential bug and security hotspots

## [4.2.2] - 2022.05.30
### Fixed
- Beacon createDateTime collected from config file or from events collection

## [4.2.1] - 2022.05.30
### Fixed
- Modified the Beacon API version from `1.0` to `1.0.1`

## [4.2] - 2022.05.23
### Added
- Explicit `/apiv1.0/info` endpoint
### Fixed
- Response errors compliant with API
- Validation errors against GA4GH's OpenAPI specification and CSCfi JSON schemas
- Make sure that dataset version created via API is of type string

## [4.1] - 2022.04.20
### Added
- Load genes when initializing demo application
- Instructions on how to load/update genes in database to filter uploaded VCFs by genes
- Added an add_dataset endpoint to create or update datasets via the API
- Return error if provided VCF path is not found before extracting genomic intervals of interest
### Fixed
- Fixed code to remove all deprecation warnings
- Dockerfile worker owns the workdir to prevent permission denied when writing temp filtered VCF file
- Adding and removing variants in background after API calls
### Changed
- `add` and `delete` endpoints return response code 202 (accepted) instead of 200 (success)

## [4.0] - 2022.03.21
### Added
- Redirect to query form page when user lands on / endpoint
- Software version on query form page
- Database name and general stats on landing page
- Tooltips on query form explaining that coordinates are 0-based
- Clicking on number of datasets in query form expands dataset description window
- Codecov config file
- Push gunicorn server image to Docker Hub on new release event
### Changed
- / endpoint redirects to info endpoint returning json beacon info
- Refactored HTML query form to be more user friendly
- Refactored query form page to better display dataset-specific results
- Refactored cli options and params to be compliant with the GNU coding standards
- Moved code checking parameters used to create a new dataset in a dedicated function under utils/add.py
- Removed optional consent codes when creating a new dataset, as they are not required and complicate code
### Fixed
- Return dataset-specific info even if allele is not found
- Instructions on how to deploy the beacon with podman as a systemd service
- Updated Black version used in GitHub action's to 22.3.0, to fix linting error in CI due to new Click release
- Dataset version is saved as a string

## [3.3.1] - 2022.03.11
### Fixed
- Long description error in setup.py file causing an error on twine before PyPI push
- Switch to codecov action v.2, since v.1 is deprecated

## [3.3] - 2022.03.11
### Added
- Push to PyPI when a new release is created
- Dockerfile-server to serve the prod app via gunicorn
- Notify admins via email when app crashes
- Queries using fuzzy string searching (allow Ns in search pattern)
- Return type annotation to all python functions
### Changed
- Install required libs from requirements.txt directly in setup.py
- Dockerfile-server image pushed to Docker Hub (cgbeacon2-server-stage) when a PR is opened or updated
- Refactored code in server/blueprints/api_v1/controllers.py
### Fixed
- Error returned when query contains non-existent dataset IDs
- Serve images as static files otherwise nginx doesn't show them

## [3.2] - 2022.02.01
### Added
- Vulture GitHub action to find unused code with 80% confidence
### Fixed
- Freeze PyMongo lib to version<4.0 to keep supporting previous MongoDB versions
### Changed
- Scan only changed .py files with Vulture
- Multi-stage Dockerfile for smaller and faster-built image


## [3.1] - 2021.10.18
### Added
- Option to provide a custom token when creating a database user
- Example of URI for connecting to a MongoDB replica set
### Fixed
- Fixed the command that runs the web app service on docker-compose
### Changed
- Replaced the alpine-based MongoDB image used in docker-compose with the official MongodDB image (db version 4.4.9)
- Launch the demo app on port 6000 when it's started via docker-compose
- Build Docker image from local repo when using docker-compose
- Use database port 27013 in docker-compose

## [3.0] - 2021.06.30
### Added
- Instructions to run the Beacon cli and web server connected to mongo as a service via Podman
- Inclusive-language check using [woke](https://github.com/get-woke/woke) github action
- Use official Docker github actions to push Docker images to Docker Hub -> Master branch (on release event) and PR branch (on push event)
- Create a demo user with a token for using the API when creating a demo instance
- Github action to deploy docs after a new release is published
### Fixed
- Links in docs pages
- Added instructions on how to install bedtools in documentation and readme file
- Build badge link in README page
- Fixed test no longer working after the release of Flask 2.0
- Replaced old docs link www.clinicalgenomics.se/cgbeacon2 with new https://clinical-genomics.github.io/cgbeacon2
- Improved code according to codefactor and Flake8 suggestions
- Indented code in a github action
- Link to docs in the README page
### Changed
- Switch to codecov in gihub actions
- Switched coveralls badge with codecov badge
- id param name in create user and create database cli command
- Use list_collection_names() instead of collection_names() that was deprecated in pymongo 3.7

cgbeacon2/models/beacon.py
# -*- coding: utf-8 -*-
import datetime

import pymongo
from cgbeacon2 import __version__

API_VERSION = "v1.0.1"

# MAP dataset internal keys to the keys expected in responses
DATASET_MAPPING = {
    "id": "_id",
    "name": "name",
    "variantCount": "variant_count",
    "callCount": "allele_count",
    "assemblyId": "assembly_id",
    "createDateTime": "created",
    "updateDateTime": "updated",
    "version": "version",
}


class Beacon:
    """Represents a general beacon object"""

    def __init__(self, conf_obj, database=None) -> None:
        self.apiVersion = API_VERSION
        self.createDateTime = conf_obj.get("createDateTime") or self._date_event(
            database, pymongo.ASCENDING
        )
        self.updateDateTime = self._date_event(database, pymongo.DESCENDING)
        self.description = conf_obj.get("description")
        self.id = conf_obj.get("id")
        self.name = conf_obj.get("name")
        self.organization = conf_obj.get("organization")
        self.sampleAlleleRequests = self._sample_allele_requests()
        self.version = f"v{__version__}"
        self.welcomeUrl = conf_obj.get("welcomeUrl")
        self.alternativeUrl = conf_obj.get("alternativeUrl")
        self.datasets = self._datasets(database)
        self.datasets_by_auth_level = self._datasets_by_access_level(database)

    def _date_event(self, database, ordering) -> datetime.datetime:
        """Return the date of the first event event created for this beacon

        Accepts:
            database(pymongo.database.Database)
            ordering(pymongo.ASCENDING or pymongo.DESCENDING)

        Returns
            event.created(datetime.datetime): date of creation of the event
        """
        if database is not None:
            events = database["event"].find().sort([("created", ordering)]).limit(1)
            for event in events:
                return event.get("created")

    def info(self) -> dict:
        """Returns a the description of this beacon, with the fields required by the / endpoint"""
        beacon_obj = self.__dict__
        beacon_obj.pop("datasets_by_auth_level")
        return beacon_obj

    def _datasets(self, database) -> list:
        """Retrieve all datasets associated to this Beacon

        Accepts:
            database(pymongo.database.Database)
        Returns:
            datasets(list)
        """
        if database is None:
            return []
        datasets = []

        for db_ds in database["dataset"].find():
            if db_ds.get("samples") is None:
                continue
            ds = {"sampleCount": len(db_ds.get("samples"))}
            for key, db_key in DATASET_MAPPING.items():
                ds[key] = db_ds.get(db_key)

            datasets.append(ds)

        return datasets

    def _datasets_by_access_level(self, database) -> dict:
        """Retrieve all datasets associated to this Beacon, by access level

        Accepts:
            database(pymongo.database.Database)
        Returns:
            datasets_by_level(dict): the keys are "public", "registered", "controlled"
        """
        datasets_by_level = dict(public={}, registered={}, controlled={})

        if database is None:
            return datasets_by_level

        datasets = database["dataset"].find()
        for ds in list(datasets):
            # add dataset as id=dataset_id, value=dataset to the dataset category
            datasets_by_level[ds["authlevel"]][ds["_id"]] = ds

        return datasets_by_level

    def _sample_allele_requests(self) -> list:
        """Returns a list of example allele requests"""

        examples = [
            {
                "alternateBases": "A",
                "referenceBases": "C",
                "referenceName": "1",
                "start": 156146085,
                "assemblyId": "GRCh37",
                "datasetIds": ["test_public"],
                "includeDatasetResponses": "HIT",
            },
            {
                "variantType": "DUP",
                "referenceBases": "C",
                "referenceName": "20",
                "start": 54963148,
                "assemblyId": "GRCh37",
                "includeDatasetResponses": "ALL",
            },
        ]
        return examples
